File: TARS/graphs/agent.py
import logging
import os
from dotenv import load_dotenv
from langchain.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import AIMessage, HumanMessage
from langchain.agents import AgentExecutor, load_tools, create_tool_calling_agent
from langchain.tools import YahooFinanceNewsTool, YouTubeSearchTool
from langchain.agents.agent_toolkits import SlackToolkit

from config.config import llm_settings
from models.generic_llms.openai import openai_llm
from models.generic_llms.anthropic import anthropic_llm
from models.generic_llms.google_ai import google_llm

from .custom_tools import get_word_length, handle_all_unread_gmail, fetch_emails_by_sender_name, read_image_tool, fetch_dms_last_x_hours, fetch_calendar_events_for_x_days

class AgentManager:
    def __init__(self):
        logging.basicConfig(level=logging.INFO)
        load_dotenv()
        llm_type = llm_settings.llm_type
        if llm_type == 'openai':
            self.llm = openai_llm
        elif llm_type == 'google':
            self.llm = google_llm
        elif llm_type == 'anthropic':
            self.llm = anthropic_llm
        else:
            raise ValueError(f"Unsupported LLM type: {llm_type}")
        self.tools = self.load_all_tools()
        self.llm_with_tools = self.llm.bind_tools(self.tools)
        self.prompt = self.define_prompt()
        self.agent = create_tool_calling_agent(self.llm, self.tools, self.prompt)
        self.agent_executor = AgentExecutor(agent=self.agent, tools=self.tools, verbose=True)

    def load_all_tools(self):
        custom_tools = [get_word_length, handle_all_unread_gmail, read_image_tool, fetch_dms_last_x_hours, fetch_calendar_events_for_x_days, fetch_emails_by_sender_name]
        request_tools = load_tools(["requests_all"])
        weather_tools = load_tools(["openweathermap-api"])
        finance_tools = [YahooFinanceNewsTool()]
        slack_tools = SlackToolkit().get_tools()
        youtube_tools = [YouTubeSearchTool()]
        return custom_tools + request_tools + weather_tools + finance_tools + slack_tools + youtube_tools

    def define_prompt(self):
        return ChatPromptTemplate.from_messages([
            ("system", "You are a helpful AI assistant named TARS. You have a geeky, clever, sarcastic, and edgy sense of humor. You were created by GP to help me be more efficient."),
            MessagesPlaceholder(variable_name="chat_history"),
            ("user", "{input}"),
            MessagesPlaceholder(variable_name="agent_scratchpad"),
        ])

    def process_user_task(self, user_task, chat_history=None):
        if user_task is None:
            raise ValueError("user_task cannot be None")
        try:
            if chat_history is None:
                chat_history = []
            invoke_params = {"input": user_task, "chat_history": chat_history}
            result = self.agent_executor.invoke(invoke_params)
            chat_history.extend([
                HumanMessage(content=user_task),
                AIMessage(content=result["output"]),
            ])
            return result["output"]
        except Exception as e:
            logging.error(f"Error in process_user_task: {e}")
            return "An error occurred while processing the task."

File: TARS/graphs/agent.js
const { loadTools, createToolCallingAgent, AgentExecutor } = require('langchain/agents');
const { ChatPromptTemplate, MessagesPlaceholder } = require('langchain/prompts');
const { AIMessage, HumanMessage } = require('langchain_core/messages');
const { YahooFinanceNewsTool, YouTubeSearchTool } = require('langchain/tools');
const { SlackToolkit } = require('langchain/agents/agent_toolkits');
const { getWordLength, handleAllUnreadGmail, fetchEmailsBySenderName, readImageTool, fetchDmsLastXHours, fetchCalendarEventsForXDays } = require('./custom_tools');
const { openaiLlm } = require('../models/generic_llms/openai');
const { googleLlm } = require('../models/generic_llms/google_ai');
const { anthropicLlm } = require('../models/generic_llms/anthropic');
const { llmSettings } = require('../config/config');

class AgentManager {
    constructor() {
        this.llm = this.initializeLlm();
        this.tools = this.loadAllTools();
        this.llmWithTools = this.llm.bindTools(this.tools);
        this.prompt = this.definePrompt();
        this.agent = createToolCallingAgent(this.llm, this.tools, this.prompt);
        this.agentExecutor = new AgentExecutor({ agent: this.agent, tools: this.tools, verbose: true });
    }

    initializeLlm() {
        const llmType = llmSettings.llmType;
        switch (llmType) {
            case 'openai':
                return openaiLlm;
            case 'google':
                return googleLlm;
            case 'anthropic':
                return anthropicLlm;
            default:
                throw new Error(`Unsupported LLM type: ${llmType}`);
        }
    }

    loadAllTools() {
        const customTools = [getWordLength, handleAllUnreadGmail, readImageTool, fetchDmsLastXHours, fetchCalendarEventsForXDays, fetchEmailsBySenderName];
        const requestTools = loadTools(["requests_all"]);
        const weatherTools = loadTools(["openweathermap-api"]);
        const financeTools = [new YahooFinanceNewsTool()];
        const slackTools = new SlackToolkit().getTools();
        const youtubeTools = [new YouTubeSearchTool()];
        return [...customTools, ...requestTools, ...weatherTools, ...financeTools, ...slackTools, ...youtubeTools];
    }

    definePrompt() {
        return new ChatPromptTemplate({
            messages: [
                { role: "system", content: "You are a helpful AI assistant named TARS. You have a geeky, clever, sarcastic, and edgy sense of humor. You were created by GP to help me be more efficient." },
                new MessagesPlaceholder({ variableName: "chat_history" }),
                { role: "user", content: "{input}" },
                new MessagesPlaceholder({ variableName: "agent_scratchpad" }),
            ]
        });
    }

    async processUserTask(userTask, chatHistory = []) {
        if (!userTask) {
            throw new Error("user_task cannot be None");
        }
        try {
            const invokeParams = { input: userTask, chat_history: chatHistory };
            const result = await this.agentExecutor.invoke(invokeParams);
            chatHistory.push(new HumanMessage({ content: userTask }), new AIMessage({ content: result.output }));
            return result.output;
        } catch (error) {
            console.error(`Error in processUserTask: ${error}`);
            return "An error occurred while processing the task.";
        }
    }
}

module.exports = AgentManager;
